src/chips/add_carry_v1.rs

```diff
@@ -0,0 +1,162 @@
+use halo2_proofs::{circuit::*, halo2curves::pasta::Fp, plonk::*, poly::Rotation};
+
+#[derive(Debug, Clone)]
+pub struct AddCarryConfig {
+    pub advice: [Column<Advice>; 3],
+    pub instance: Column<Instance>,
+    pub selector: Selector,
+}
+
+#[derive(Debug, Clone)]
+pub struct AddCarryChip {
+    config: AddCarryConfig,
+}
+
+impl AddCarryChip {
+    pub fn construct(config: AddCarryConfig) -> Self {
+        Self { config }
+    }
+
+    pub fn configure(
+        meta: &mut ConstraintSystem<Fp>,
+        advice: [Column<Advice>; 3],
+        selector: Selector,
+        instance: Column<Instance>,
+    ) -> AddCarryConfig {
+        let col_a = advice[0];
+        let col_b = advice[1];
+        let col_c = advice[2];
+        let add_carry_selector = selector;
+
+        // Enable equality on the advice and instance column to enable permutation check
+        meta.enable_equality(col_a);
+        meta.enable_equality(col_b);
+        meta.enable_equality(col_c);
+        meta.enable_equality(instance);
+
+        // enforce dummy hash function by creating a custom gate
+        meta.create_gate("accumulate constraint", |meta| {
+            let s = meta.query_selector(add_carry_selector);
+            let prev_b = meta.query_advice(col_b, Rotation::prev());
+            let prev_c = meta.query_advice(col_c, Rotation::prev());
+            let a = meta.query_advice(col_a, Rotation::cur());
+            let b = meta.query_advice(col_b, Rotation::cur());
+            let c = meta.query_advice(col_c, Rotation::cur());
+
+            // Previous accumulator amount + new value from a_cell
+            // using binary expression (x_n-4 * 2^16) + (x_n-3 * 2^8) + ... + (x_n * 2)
+            vec![
+                s * ((a + (prev_b * Expression::Constant(Fp::from(1 << 16))) + prev_c)
+                    - ((b * Expression::Constant(Fp::from(1 << 16))) + c)),
+            ]
+        });
+
+        AddCarryConfig {
+            advice: [col_a, col_b, col_c],
+            instance,
+            selector: add_carry_selector,
+        }
+    }
+
+    // Initial accumulator values from instance for expreiment
+    pub fn assign_first_row(
```

Comment:
I think that the first value assigned to the advice table should not be copied from the instance column, as there might be cases where this value is not public. 

But since we remove the `assign_advice_from_instance` we now need to constrain the first value to be equal to its 16_bits chunks decomposition. 

I suggest refactoring it as follows:

- `assign_first_row`only assign 0 to `acc_hi` and 0 to `acc_lo`

In the circuit `synthesize()` pass the first value to the `assign_advice_row` such that the constraint already in place checks that `value + prev_acc_hi + prev_acc_lo + cur_acc_hi + cur_acc_low = 0` which is equal to `value + 0 + 0 + cur_acc_hi + cur_acc_low = 0`  which is the constraint the want to set for our first value

Reply:
Yes, I wanted to use the `assign_advice_from_constant` instead of, but I could not resolve the error `NotEnoughColumnsForConstants` when I wrote this code. Now I realized that need at least one Fixed column for that.

Refactored at a2e47a0

Reply:
Perfect! Added a test at 0e6e100 while trying the chip.